Indexer: Do not create autodocs entries unless addon-docs installed

ndelangen opened the pull request:
Closes https://github.com/storybookjs/storybook/issues/31326

## What I did

This pull request includes updates to streamline the handling of documentation options and improve type definitions in the codebase. The most significant changes involve refactoring the `docs` preset logic, enhancing type safety, and adding conditional checks for autodocs.

### Improvements to documentation handling:

* Refactored the `docs` preset in `code/addons/docs/src/preset.ts` to use the `PresetProperty<'docs'>` type and added support for passing `docsMode` from `CLIOptions`. This ensures better type safety and flexibility in handling documentation options.
* Removed the redundant `docs` export from `code/core/src/core-server/presets/common-preset.ts`, consolidating the logic into the updated implementation in `preset.ts`.

### Type definition enhancements:

* Updated imports in `code/addons/docs/src/preset.ts` to include additional types like `CLIOptions` and `StorybookConfigRaw`, ensuring more precise type usage throughout the file.

### Conditional autodocs generation:

* Modified the `StoryIndexGenerator` in `code/core/src/core-server/utils/StoryIndexGenerator.ts` to only create documentation entries when autodocs are enabled and the `docs` option is defined. This adds a safeguard against unintended behavior.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-31331-sha-87ed948b`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-31331-sha-87ed948b sandbox` or in an existing project with `npx storybook@0.0.0-pr-31331-sha-87ed948b upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-31331-sha-87ed948b`](https://npmjs.com/package/storybook/v/0.0.0-pr-31331-sha-87ed948b) |
| **Triggered by** | @shilman |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`norbert/docs-preset-property-indexer-fix`](https://github.com/storybookjs/storybook/tree/norbert/docs-preset-property-indexer-fix) |
| **Commit** | [`87ed948b`](https://github.com/storybookjs/storybook/commit/87ed948b6e00bb75f8780bf352d9a449493651ea) |
| **Datetime** | Fri May  2 07:01:25 UTC 2025 (`1746169286`) |
| **Workflow run** | [14790433123](https://github.com/storybookjs/storybook/actions/runs/14790433123) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=31331`_
</details>
<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

This PR fixes a bug where Storybook was creating documentation entries even when the addon-docs package wasn't installed, leading to runtime errors.

- Modified `StoryIndexGenerator.ts` to only create docs entries when both autodocs tag and docs option are present
- Removed redundant `docs` export from `common-preset.ts` to consolidate logic in addon-docs package
- Updated `preset.ts` to handle `docsMode` conditionally based on both `docsOptions` and CLI settings
- Fixed issue where minimal Storybook installations were incorrectly showing docs entries in sidebar



<!-- /greptile_comment -->

ndelangen committed:
Move the docs preset property handling to addon docs, so that addon docs can be easily detected in the indexer, do not generate any docs entries when addon-docs is not enabled

ndelangen committed:
Remove unused import of DocsOptions from preset.ts

shilman commented on code/addons/docs/src/preset.ts:
```diff
@@ -129,12 +134,15 @@ async function webpack(
-    autodocs: 'tag',
   };
+
+  if (docsOptions && docsMode) {
```
Why do we need this?

ndelangen replied:
That was already there, I just refactored it.

The big difference is that now I detect addon-docs, by the presence of options.docs, which used to be set by common-preset.

By moving it to addon-docs itself, now `options.docs` becomes `undefined` when `addon-docs` is NOTv set up.

shilman reviewed (commented):
Looks good to me, but I don't understand how docsMode  relates to anything.

ndelangen committed:
Merge branch 'next' into norbert/docs-preset-property-indexer-fix

ndelangen committed:
Merge branch 'next' into norbert/docs-preset-property-indexer-fix

ndelangen committed:
Refactor docs handling in Storybook configuration and related files

- Updated import path for BROWSER_TARGETS in Storybook main configuration.
- Commented out unused docs addon in Storybook configuration.
- Modified docs preset to handle options more flexibly and removed unnecessary checks.
- Adjusted build index and static functions to apply docs options without defaults.
- Cleaned up unused docs preset code and ensured proper handling of docs options in StoryIndexGenerator.
- Added console logs for debugging docs options during build processes.

ndelangen committed:
Remove console logs related to docs options in build index and static functions

ndelangen commented:
I tested this locally, and it works now!
@shilman please review?

ndelangen committed:
Merge branch 'next' into norbert/docs-preset-property-indexer-fix

ndelangen committed:
Enhance Storybook configuration by enabling docs addon and adding docs stories directory

- Re-enabled the '@storybook/addon-docs' in the Storybook main configuration.
- Added a new directory for docs stories to improve documentation handling.
